### DISCUSSION TITLE
Event data caching issues

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#9853] [General]
## Scope

Events, `Canvas#_cacheTransformEventData`, `Canvas#_resetTransformEventData`

Issues regarding the caching of:
- _pointer
- _absolutePointer
- _target

In the events: down, move, up, dblclick, wheel, drag & drop, contextmenu

During the lifecycle of an event handler in fabricJS.
For event handler we mean the top level functions that wrap a lot of code and logic between a call of `_cacheTransformEventData` and a cleanup call of `_resetTransformEventData`.

The cache is accessed implicitly through other methods.
For `_pointer` or `viewportPoint` the cache is accessed through `getViewportPoint`.
For `_absolutePointer` or `scenePoint` the cache is accessed through `getScenePoint`.
When those functions are called between a call of `_cacheTransformEventData` and `_resetTransformEventData` they will short circuit and return the stored values.

_target is on its own is accessed directly from `_handleEvent` to create the event context returned to the user.
_target is equally affected because findTarget function will rely on cached `getViewportPoint` and `getViewportPoint` so all of the values suffers from the same weakness

Issues may arise when a developer inject code using events handlers (`.on('mouse:move')`) inside the fabricJS event flow that require fresh calculation of those values. Fresh calculation won't happen

## Why does fabric use event caching?

>  the reason why those values were cached is because traversing the dom was expensive.
In a mouse move situation was painful in the moment we started to fire more events or check more often the pointer

> _Originally posted by @asturur in https://github.com/fabricjs/fabric.js/issues/9850#issuecomment-2094534183_

## Why do I think fabric uses event caching?

Fabric creates event context repeatedly when it fires events and uses that context in many cases.
However, the main canvas methods do not use it and leaf consumers at the very edge do not, e.g.
https://github.com/fabricjs/fabric.js/blob/a055616e113dffb4be326f426f9dff177e86b5cb/src/shapes/IText/ITextClickBehavior.ts#L246-L247

It seems that it is not used for historical reasons. The code was written and wasn't refactored to limit changes.
I would argue that caching was done **ONLY** because the context didn't reach all consumers. So instead of breaking signatures of those consumers in order to provide better context, caching was introduced.

## List of issues caused by the current caching approach

- #8335: if an object is being moved/transformed by the user and during is discarded programmatically the object will continue moving and will be inactive only on mouseup. We exposed `endCurrentTransform` to fix more related issues IIRC.
- #7713 changing vpt during selection makes the object jump in space
- change of viewport during any event cached event
- change of canvas position/size/css affecting size props in the dom during any cached event

### My Motivation

I prefer refactoring than workarounds.

I hate gotchas and even more fabric gotchas.
They are time sinks.
I have been working with fabric for a number of years as a contributor and before that as a consumer. 
The fact that I still bump into gotchas makes me 😡.
It states to me that fabric is immature for PROD.

I want to fix this issue without degrading performance.
This means refactoring the code.
            
### A Different Design: context vs. caching

Providing context is the better way because:
It is stateless and declarative.
It is easy to test, maintain and reason about.
It is easy to override, subclass and customize. e.g. a dev could control the context and pass whatever they want as long as it follows the context convention and control fabric with ease.

Fabric creates event context repeatedly when it fires events and uses that context in many cases.
Fabric should use that context and declare it at the beginning of each event handler and pass it down to where it didn't reach until now.
This means that instead of recreating the context in `_handleEvent` we pass it down from the caller (the top level event hander) and use in the caller as well.
To do that properly we must break the drag and drop API, `handleMultiselection`, `TextEditingManager` (which is private and not exported I think) and a number of text methods by changing their signatures to accept the context including the pointer data.

> I have worked on removing event caching: https://github.com/fabricjs/fabric.js/compare/fix/rm-event-caching?expand=1

> _Originally posted by @ShaMan123 in https://github.com/fabricjs/fabric.js/issues/9850#issuecomment-2094841735_
            

#### Document how to handle mouse events and caching

Don't change anything but clearly state that if you need to change shape, position or viewport of the canvas during a mouse event handler, you have to either use a defer-like statement or you need to call `_cacheTransformEventData` before trying to calculate new  values for those mentioned functions

#### Dismiss cached values before event handling

> We could just reset the cache before firing the event probably and fire the event as late as it makes sense.

At the time of giving the event context to the user, in handle event, fabricJS saved all the time ti could save with this cache and can now remove the cached values.
Those would need to be recreated after the event is handled for all the 'before:*' events

#### Remove caching

> we could just stop caching the one affected by the viewport port so that is just a matrix multiplication and is not expensive.
> Meaning that the actual point you click on the canvasElement the non scene one is not going to change, and that is the expensive one to calculate.

> _Originally posted by @asturur in https://github.com/fabricjs/fabric.js/issues/9850#issuecomment-2094534183_

Though this sounds legit as a fix for the mentioned bug in #9850 (changing viewport transform in an event handler and calling `Canvas#getScenePoint`) it doesn't fix other bugs caused by the same issue.

Consider a RTL css layout app with the following case:
During an event, when caching is active, an event handler triggers a canvas resize.
Since pointers are measured from the left, the viewport point will also become invalid, making the proposal to keep part of pointer caching as is invalid.

> Is non breaking anyway is just performances

> _Originally posted by @asturur in https://github.com/fabricjs/fabric.js/issues/9850#issuecomment-2094534183_

I do not want to degrade performance so removing caching entirely sounds bad if indeed the dom calculations are heavy.
However, I do prefer something correct than something that is fast and is correct most of the time. That is the exact cause of a gotcha.
Regardless, PROD accepts only something that is correct and fast.
So if fabric is meant for PROD removing caching entirely is not legit.

#### Assign the cached values to the DOM event itself

> If the perf opt is really that significant (and that is not clear) then we can assign the values to the pointer event so that `e.scenePoint` could be accessed instead of calling `getScenePoint`.

> Originally posted by @ShaMan123 in https://github.com/fabricjs/fabric.js/pull/9850#issue-2279209833

The event will be assigned with the relevant values and the dev can choose to opt in to use them by replacing the calls to `getViewportPoint` `getScenePoint` to increase performance, there will be no caching by fabric. 
Since `getViewportPoint` `getScenePoint` are new methods introduced in v6 and require migration deprecating  v5 `getPointer`, we can offer a more clever way to migrate. We can add a section about this in the migration guide or in #8299  without impl this while leaving it to the dev to use the event context from the event handler.
This is valid and does not break anything apart for a ton of tests.
This solution also follows event conventions, meaning, that the data is correct at the moment of the event being fired. If something changed during the time that has passed from firing the event, the dev understands that they can't trust the values any longer and resort to calling `getScenePoint` or `getViewportPoint`.

_Originally posted by @ShaMan123 in https://github.com/fabricjs/fabric.js/issues/9850#issuecomment-2095186147_

**Comment by ShaMan123** [comment]
RESERVED for edits on the OP
@jiayihu 

Edit: @asturur added introduction of what is event data caching in fabricJS
edit: @asturur added more specific of where the cache is today
edit: @asturur  added the document only option and divided remove caching in remove caching and Dismiss cached values before event handling

**Reply by asturur** [reply]
What it is the difference between `Why does fabric use event caching?` and `Why do I think fabric uses event caching?` Can you clarify and edit accordingly? If there is a difference in why is used and why you think is used we need to adjust that.

**Reply by asturur** [reply]
Please describe your work on remove event caching that you made in the branch in english, if i have to understand it from code changes is a PR review.
Otherwise you end up having to make a branch of every possible solution rather than discussing it

**Reply by asturur** [reply]
Can you please list pedantically all the bugs you can think of as you mentioned the other 2, in a list:
Possible bugs:
- change of viewport during any mouse event
- change of canvas location in the dom during any mouse event
- change of canvas size in the dom during any mouse event
etc.. etc...

**Reply by asturur** [reply]
Can you describe how 'Assign the cached values to the DOM event itself' would interact with the current issue of getViewportPoint and getScenePoint? those functions will still try to use the cached value appended to the event object and so the user would still need to know that there is a cache to manually invalidate. This solution changes the location of the cache, from the canvas to the event, and saves us from the manual cleanup, but how does change the the implementation in a meaningful way?

**Reply by ShaMan123** [reply]
Nope.
There will be no caching on fabric. The event will be assigned values and the dev can choose to opt in to use them by replacing the calls to `getViewportPoint` `getScenePoint` to increase performance.
As I mentioned, since those are new methods and we have deprecated `getPointer` the dev must migrate anyways so we offer a more clever way to migrate. We can add a section about this in the migration guide or in the breaking changes issue without impl this and leave it to the dev to use the event context from the event handler.

**Reply by ShaMan123** [reply]
aded to description

**Reply by asturur** [reply]
ok so that solution there is remove caching. period. So is one together with the other remove caching

**Reply by ShaMan123** [reply]
A different suggestion that keeps perf intact

**Reply by ShaMan123** [reply]
Will add about the PR now

**Reply by ShaMan123** [reply]
Please read the part about the PR and comment if not good enough

**Reply by asturur** [reply]
> A different suggestion that keeps perf intact
Is really unclear, i ll reread again later when you are done. so far 'Assign the cached values to the DOM event itself' is 'remove caching' as i understand it.

**Reply by ShaMan123** [reply]
Maybe related https://github.com/fabricjs/fabric.js/pull/7713

**Comment by ShaMan123** [comment]
related #8335

**Reply by asturur** [reply]
If it is related to 8335 pull in the relevant related concepts in the document

**Reply by ShaMan123** [reply]
Added
